alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+# TODO - Need to modularise torch and tensorflow imports and use. e.g. has_tensorflow and has_pytorch etc
+from alibi_detect.utils.registry import registry
+from alibi_detect.utils.schemas import SupportedModels  # type: ignore[attr-defined]
+from alibi_detect.utils.validate import validate_config
+from alibi_detect.utils.tensorflow._loading import load_model as load_model_tf, \
+    prep_model_and_emb as prep_model_and_emb_tf, load_kernel_config as load_kernel_config_tf, \
+    load_optimizer as load_optimizer_tf, load_embedding as load_embedding_tf, load_detector_legacy, Detectors
+#  from alibi_detect.utils.pytorch._loading import set_device
+import tensorflow as tf  # TODO - this is currently only required for FIELDS_TO_DTYPE conversion. Remove in future.
+from transformers import AutoTokenizer
+import numpy as np
+from typing import Union, Optional, Callable, Any
+from copy import deepcopy
+from functools import partial
+from pathlib import Path
+import dill
+import os
+import toml
+from importlib import import_module
+import logging
+
+logger = logging.getLogger(__name__)
+
+REQUIRES_BACKEND = [
+    'ClassifierDrift',
+    'ClassifierUncertaintyDrift',
+    'LearnedKernelDrift',
+    'LSDDDrift',
+    'MMDDrift',
+    'RegressorUncertaintyDrift',
+    'SpotTheDiffDrift'
+]
+
+# Fields to resolve in resolve_cfg ("resolve" meaning either load local artefact or resolve @registry, conversion to
+# tuple, np.ndarray and np.dtype are dealt with separately).
+FIELDS_TO_RESOLVE = [
+    ['preprocess_fn'],
+    ['preprocess_fn', 'src'],
+    ['preprocess_fn', 'model'],
+    ['preprocess_fn', 'embedding'],
+    ['preprocess_fn', 'tokenizer'],
+    ['x_ref'],
+    ['model'],
+    ['optimizer'],
+    ['reg_loss_fn'],
+    ['kernel'],
+    ['dataset'],
+    ['kernel', 'src'],
+    ['kernel', 'proj'],
+    ['kernel', 'kernel_a', 'src'],
+    ['kernel', 'kernel_b', 'src'],
+    ['initial_diffs']
+]
+
+# Directories to amend before resolving config (fields to prepend config file dir to)
+DIR_FIELDS = [
+    ['preprocess_fn'],
+    ['preprocess_fn', 'src'],
+    ['preprocess_fn', 'model', 'src'],
+    ['preprocess_fn', 'embedding', 'src'],
+    ['preprocess_fn', 'tokenizer', 'src'],
+    ['x_ref'],
+    ['model', 'src'],
+    ['kernel'],
+    ['kernel', 'src'],
+    ['optimizer'],
+    ['reg_loss_fn'],
+    ['kernel', 'proj', 'src'],
+    ['kernel', 'kernel_a', 'src'],
+    ['kernel', 'kernel_b', 'src'],
+    ['initial_diffs']
+]
+
+# Fields to convert from list to tuple in resolve_cfg
+FIELDS_TO_TUPLE = [
+    ['detector', 'kwargs', 'input_shape']
+]
+
+# Fields to convert from list to np.ndarray in resolve_cfg
+FIELDS_TO_ARRAY = [
+    ['sigma'],
+    ['kernel', 'sigma'],
+    ['kernel', 'kernel_a', 'sigma'],
+    ['kernel', 'kernel_b', 'sigma'],
+]
+
+# Fields to convert from str to np.dtype
+FIELDS_TO_DTYPE = [
+    ['preprocess_fn', 'dtype']
+]
+
+
+def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Detectors:
```

Comment:
Can we document the use cases for supplying `kwargs` here?

Reply:
This has actually made me realise we (I) might need to think again about the `custom_objects` field. The `kwargs` is passed to `load_detector_legacy`, which accepted additional generic kwarg's for a number of things. The main one of interest is that if `custom_objects` was in `kwargs`, it would be passed to `load_model`:

https://github.com/SeldonIO/alibi-detect/blob/fc83c707d359d4e5009b7cfb84191748a5625bdd/alibi_detect/utils/tensorflow/_loading.py#L310-L311

So actually what I currently have where `custom_object` is read/written to `config.toml` perhaps doesn't make sense, as the existing functionality is actually to pass in as kwarg to `load_detector`. But we could support spec'ing it via config as a pre-registered function/class... but then I'm wondering if it makes more sense to explore how well the existing tensorflow custom object registering works here https://www.tensorflow.org/api_docs/python/tf/keras/utils/register_keras_serializable.

TL;DR I'm wondering if we should just remove `custom_object` support entirely for now, and add back in as a later addition?

Reply:
`custom_objects` would only be used if it needed to be passed to the tensorflow load function. We currently only do this for our `resnet56` model (see https://github.com/SeldonIO/alibi-detect/pull/469#issuecomment-1080434048).

If we removed I can't see there being many issues in the short-term, except that we are technically reducing capability (as the legacy `load_detector` accepted `custom_objects` as a kwarg).